scripts: zephyr_module: Add CPE support (#66495)
zephyrproject-rtos/zephyr: doc/develop/modules.rst

   use ``bar`` as its module name if nothing is specified in
   :file:`zephyr/module.yml`.

Module version
==============

> Each Zephyr module should have a version number to improve tracking.

Reviewer: `should` seems a bit strong in this case, especially as it would indicate that modules are having some form of release cycles. This may be true for some modules, but other modules revision might be considered to be identified (thus indirectly follow the Zephyr version) through the SHA in the manifest, which again is identified by the Zephyr revision / version.

```suggestion
Each Zephyr module may have a version number to improve tracking.
```

This allows control of the build inclusion to be described externally to the
Zephyr module.

> SBOM integration

Reviewer: Should we turn this around.

Instead of having `SBOM integration`, then describe what content is (example: `Vulnerability monitoring`), and then instead describe that this information will be included in the sbom.

The vulnerability monitoring / information can be of interest, even without an SBOM.

(Just a thought, not a request for a change)

====================

The module description file :file:`zephyr/module.yml` can be used to improve the
generated SBOM and vulnerability monitoring.

> The version and the CPE associated to the module can be provided using

Reviewer: when using abbreviations like: `CPE` and `CVE` (and soon `VEX`), we should write them in full once, so that even users unfamiliar with the terms know what they mean.

``version`` and ``cpe`` fields:

.. code-block:: yaml

>    version: <module-version>

Reviewer: Let's move the `version` out of `SBOM integration`.

`version` has value on its own, and should be considered first class citizen in the module yaml.
Just like `name` is independent: https://docs.zephyrproject.org/latest/develop/modules.html#module-name

> be monitored for your module. The supported format are:

Reviewer 2: ```suggestion
be monitored for your module. The supported formats are:
```

- CPE (Common Platform Enumeration)
- PURL (Package URL)

> .. code-block:: yaml

Reviewer 2: it might be my own preference but I would also add an actual example of how things would look like with real values, ex. by using mbed TLS, or whatever, as an example. Also, as a way to make sure people instantly realize where the new section belongs, i would have the YAML snippet show more than just the relevant section, ex. include name and build entries

```
name: lvgl
build:
  ...
vulnerabilities:
   external-references:
     - ...
```

   vulnerabilities:

>      external-references:

Reviewer 2: I *think* I'd be more specific and make this two lists: `cpe-identifiers:` and `purl-identifiers:` as `external-references:` sounds really vague (unless you do want to allow people to potentially add whatever types of external-references they want, but at least now it looks like an external-identifier entry that matches neither of the two regexps gets dropped)

.. code-block:: yaml

>    vulnerabilities:

Reviewer: not sure `vulnerabilities` is the best key at this level.
It can give the impression that it will contain information regarding actual vulnerabilities, which is not the case.

You already propose a `cpe` entry, so why can we not provide the `remote` as a sub-field under there ?

Author: An evolution to `vulnerabilities` would be to add the field `vex` which would point to a vex file allowing to filter CVE.
Do you think it would fit in `cpe` field or should we find an other name (not sure what it would be though) ?

Reviewer: thanks for sharing some thoughts for future direction of this.

With a `vex` (or vdr) field, then vulnerabilities starts to make more sense as name.
I don't have a better suggestion for another name.

Reviewer 2: how about simply`security:`? `security_info:`? 
I do agree with @tejlmand that "vulnerabilities" makes it sound like it lists vulnerabilities or something

Reviewer: I like security_info more.
Broad enough to cover the vex info, while still specific enough to be useful for readers to understand the context.

This allows control of the build inclusion to be described externally to the
Zephyr module.

.. _modules-vulnerability-monitoring:

Vulnerability monitoring
========================

The module description file :file:`zephyr/module.yml` can be used to improve vulnerability monitoring.

If your module needs to track vulnerabilities using an external reference

> (e.g your module is forked from an other repository), you can use the ``security`` section.

Reviewer 2: ```suggestion
(e.g your module is forked from another repository), you can use the ``security`` section.
```

It contains the field ``external-references`` that contains a list of references that needs to
be monitored for your module. The supported formats are:

- CPE (Common Platform Enumeration)
- PURL (Package URL)

.. code-block:: yaml

   security:
     external-references:
       - <module-related-cpe>
       - <an-other-module-related-cpe>
       - <module-related-purl>

A real life example for `mbedTLS` module could look like this:

.. code-block:: yaml

   security:
     external-references:
       - cpe:2.3:a:arm:mbed_tls:3.5.2:*:*:*:*:*:*:*

>        - pkg:github/Mbed-TLS/mbedtls@V3.5.2

Reviewer 2: This actually poses a good question regarding what CPE and PURLs should be used for the Zephyr forks of the upstream project. 
Will it be expected that e.g. the https://github.com/zephyrproject-rtos/mbedtls module shall reference the CPE/PURL of the most recently released version of the upstream mbedtls project?

BTW I think it would be great if the sample code block would show a bit more of the "surroundings" so that folks have an immediate clue regarding the nesting level of the `security` key without having to look at the schema, i.e

```
name: mbedtls
build:
  ...
security:
   external-references:
       - cpe:2.3:a:arm:mbed_tls:3.5.2:*:*:*:*:*:*:*
       - pkg:github/Mbed-TLS/mbedtls@V3.5.2
```

Author: The Zephyr forks must used the CPE/PURL with the version it forked for the current version. For example, for mbedTLS, the last commit merge the upstream 3.5.2 into the zephyr fork (https://github.com/zephyrproject-rtos/mbedtls/commit/6ec4abdcda78dfc47315af568f93e5ad4398dea0) , so `module.yml` must contain the example I described.

For the example, I did not want to add keys that are not related to my modification (it would probably not be updated if a changes occurs in `build` for example). Let me know what I should do.

Reviewer 2: OK, I now understand why you wouldn't list other keys in case the schema changes, still wonder if it would make sense to at least mention something like " ... you can use the top-level ``security`` section" or something, just to help people know quicker where they need to put it

Re: CPE/PURL, part of my question was also (but really just a question for my own education and not something that should get in the way of this PR in any way) how to deal with the Zephyr module forks where it's more than just mirroring the upstream module but also including our own patches (which can come with their own vulnerabilities) on top of it. Should the "external references" list both the CPE/PURL of the upstream module + a CPE/PURL corresponding to the Zephyr fork?

Author: You are totally right. This was part of the initial discussion (#53479). At first I wanted zephyr modules maintainers to have the responsibility of the CVE for their project. We finally decided, for now, it wouldn't be the case. 
For now, CVE in zephyr modules are not reported (no CPE or via Github Advisiory Database). This is something I would like to change. 
This PR is part of multiple PR (I hope) that could lead to what you propose.

The future steps I would like to have:

 - add VEX support in `module.yml` (paths to VEX)
 - automate vulnerabilities detection from upstream (in zephyr pipeline)
 - add a CPE for each zephyr modules

But this is a discussion to have with the security team / @ceolin and see what's possible or not. Let's take this step by step :)

I hope it answers your question.

Reviewer 2: > I hope it answers your question.

it very much does :) Thanks!

Reviewer 3: Yep, ideally Zephyr forks will have their on CPE, but adding it right now would be useless since these CPEs would no match any CVE database.  From a practical point of view, what we really need now is a way to monitor CVEs in the projects we have forked.

This allows control of the build inclusion to be described externally to the
Zephyr module.

Vulnerability monitoring
========================

The module description file :file:`zephyr/module.yml` can be used for vulnerability monitoring
and improve the generated SBOM output.

> 

Reviewer: Very good.

Maybe at this location specify the `version:` field should be used.
For example:
```suggestion

When a module monitors vulnerabilities then the module should also define the `version` field, :ref:`module_version`.
```

Author: Your comment makes me think this `version` field creates confusion. There was a misunderstanding from my part in the original issue.

The goal was to provide a version for the module and help complete the SBOM for the module (and not searching vulnerabilities from the forked repository for example). The `external-references` field must be used for that (and CPE and PURL specifications allow to include the version)

I think we should remove the `version` field from this PR and I will clarify this for a next PR. Are you okay with this ?

Author: It seems that there are some clarification (raised by your comment) to be done on this PR with the security group. In the meantime, I converted it into draft and I'll update it when I have all the information I need.

I'm sorry for this.

Reviewer: if you believe it clarifies the usage and intentions of the field, then it's fine.

The clearer the intention of a PR, the better.
